### PR TITLE
WMTheme: support macOS

### PR DESCRIPTION
**Pull request by CarterLi**
Note the `AppleAccentColor` part was not tested because the `AppleAccentColor` doesn't exist in the plist file

### CODE REVIEW

File: src/modules/wmtheme.c [hunk 1]
```diff
@@ -229,6 +229,87 @@ static void printOpenbox(FFinstance* instance, const FFstrbuf* dePrettyName)
...
+        ffplist_get_uint_val(pWmThemeColor, &wmThemeColor);
+    switch (wmThemeColor)
+    {
+        case (uint64_t)-1: ffStrbufAppendS(theme, " (Graphite)");
+        case 0: ffStrbufAppendS(theme, "Red");
+        case 1: ffStrbufAppendS(theme, "Orange");
+        case 2: ffStrbufAppendS(theme, "Yellow");
+        case 3: ffStrbufAppendS(theme, "Green");
+        case 5: ffStrbufAppendS(theme, "Purple");
+        case 6: ffStrbufAppendS(theme, "Pink");
```
Comment (LinusDierheimer) [inline src/modules/wmtheme.c:268]: This needs break statements.

File: src/util/FFstrbuf.c [hunk 2]
```diff
@@ -29,6 +29,14 @@ void ffStrbufInitCopy(FFstrbuf* strbuf, const FFstrbuf* src)
     ffStrbufAppend(strbuf, src);
 }
 
+void ffStrbufInitS(FFstrbuf* strbuf, const char* str)
+{
+    uint32_t bufSize = (uint32_t)strlen(str) + 1;
+    ffStrbufInitA(strbuf, FASTFETCH_STRBUF_DEFAULT_ALLOC > bufSize ? FASTFETCH_STRBUF_DEFAULT_ALLOC : bufSize);
+    memcpy(strbuf->chars, str, bufSize);
+    strbuf->length = bufSize - 1;
```
Comment (LinusDierheimer) [inline src/util/FFstrbuf.c:37]: Is this really faster than an init and an append?

Comment (CarterLi) [reply src/util/FFstrbuf.c:37]: Ok, but isn't this a bug?

https://github.com/LinusDierheimer/fastfetch/blob/master/src/util/FFstrbuf.c#L100

Comment (CarterLi) [reply src/util/FFstrbuf.c:37]: https://github.com/LinusDierheimer/fastfetch/blob/master/src/util/FFstrbuf.c#L112

Shouldn't be `strbuf.length + length + 1`?

Comment (LinusDierheimer) [reply src/util/FFstrbuf.c]: No, `ffStrbufEnsureFrees` ensures that there is enough space to append the given length, including the null byte.

**Comment by CarterLi** [comment]
Ok, I misunderstood.

However I'd say that the existing code is not optimised.

```c
void ffStrbufAppendS(FFstrbuf* strbuf, const char* value)
{
    if(value == NULL)
        return;

    for(uint32_t i = 0; value[i] != '\0'; i++)
    {
        if(i % 16 == 0)
            ffStrbufEnsureFree(strbuf, 16);
        strbuf->chars[strbuf->length++] = value[i];
    }

    strbuf->chars[strbuf->length] = '\0';
}
```

1. Reallocating in a loop is really not a good idea because memory allocating is slow. It's always better to pre-calculate the required buffer size.
2. `memcpy` is always faster than hand coded plain loop ( if compiler failed to optimise the loop to a memcpy call ). A well coded `memcpy` ( such as the one provided in glibc ) uses [SIMD instructions](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) copies 16 ( with SSE2 ) or 32 ( with AVX ) bytes per loop, and can be about 10~20 times faster than hand coded plain loop which copies only 1 byte per loop. The larger the buffer is, the faster memcpy can be.

Quick benchmark: https://quick-bench.com/q/rv4D5N9zynZRRcWb2Td7y0BVneI

image

**Comment by LinusDierheimer** [comment]
Makes sense. Thanks for looking into this.

**Comment by CarterLi** [comment]
ffStrbufAppendS can be further optimised by make it inline-able.

```c
static inline void ffStrbufAppendS(FFstrbuf* strbuf, const char* value)
{
    if(value == NULL)
        return;

    ffStrbufAppendNS(strbuf, (uint32_t)strlen(value), value);
}
```

If the value is a string literal, its length can be known at compile time. In which case `(uint32_t)strlen(value)` can be completely optimised away.

```c
// Before compiling
ffStrbufAppendS(&strbuf, "123456789");

// After compiling
ffStrbufAppendNS(&strbuf, 9, "123456789");
```

**Comment by LinusDierheimer** [comment]
With the current implementation it might even make sense to force inline it.

**Comment by CarterLi** [comment]
Quick test: https://gcc.godbolt.org/z/z3rrn9vfa

image

**Comment by LinusDierheimer** [comment]
I don't know enough much about that, but i don't think this applies to fastfetch, because the method implementation is in another compilation unit. Might not be a problem with IPO tho. If we want to make sure it is getting inlined, we'd have to implement it as inline method in the header.

**Comment by CarterLi** [comment]
Yes. A inline-able function must be in the header. Just move them into `FFstrbuf.h` you are done.

**Pull request merged**